From 65d2c00af5c54355cc84592508d90549593159cc Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Tue, 10 Mar 2020 14:33:09 +0100
Subject: [PATCH 36/36] netlink: add netlink handler for schannels (-L)

Implement "ethtool -L <dev> ..." subcommand to set network device channel
counts using ETHTOOL_MSG_CHANNELS_SET netlink message. These are
traditionally set using ETHTOOL_SCHANNELS ioctl request.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c          |  1 +
 netlink/channels.c | 69 +++++++++++++++++++++++++++++++++++++++++++++-
 netlink/extapi.h   |  2 ++
 3 files changed, 71 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index cfc9a59f8d85..ec82ccfa1e13 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -5355,6 +5355,7 @@ static const struct option args[] = {
 	{
 		.opts	= "-L|--set-channels",
 		.func	= do_schannels,
+		.nlfunc	= nl_schannels,
 		.help	= "Set Channels",
 		.xhelp	= "               [ rx N ]\n"
 			  "               [ tx N ]\n"
diff --git a/netlink/channels.c b/netlink/channels.c
index e07e29aac897..3ee6f3beb16d 100644
--- a/netlink/channels.c
+++ b/netlink/channels.c
@@ -1,7 +1,7 @@
 /*
  * channels.c - netlink implementation of channel commands
  *
- * Implementation of "ethtool -l <dev>"
+ * Implementation of "ethtool -l <dev>" and "ethtool -L <dev> ..."
  */
 
 #include <errno.h>
@@ -11,6 +11,7 @@
 #include "../internal.h"
 #include "../common.h"
 #include "netlink.h"
+#include "parser.h"
 
 /* CHANNELS_GET */
 
@@ -61,3 +62,69 @@ int nl_gchannels(struct cmd_context *ctx)
 		return ret;
 	return nlsock_send_get_request(nlsk, channels_reply_cb);
 }
+
+/* CHANNELS_SET */
+
+static const struct param_parser schannels_params[] = {
+	{
+		.arg		= "rx",
+		.type		= ETHTOOL_A_CHANNELS_RX_COUNT,
+		.handler	= nl_parse_direct_u32,
+		.min_argc	= 1,
+	},
+	{
+		.arg		= "tx",
+		.type		= ETHTOOL_A_CHANNELS_TX_COUNT,
+		.handler	= nl_parse_direct_u32,
+		.min_argc	= 1,
+	},
+	{
+		.arg		= "other",
+		.type		= ETHTOOL_A_CHANNELS_OTHER_COUNT,
+		.handler	= nl_parse_direct_u32,
+		.min_argc	= 1,
+	},
+	{
+		.arg		= "combined",
+		.type		= ETHTOOL_A_CHANNELS_COMBINED_COUNT,
+		.handler	= nl_parse_direct_u32,
+		.min_argc	= 1,
+	},
+	{}
+};
+
+int nl_schannels(struct cmd_context *ctx)
+{
+	struct nl_context *nlctx = ctx->nlctx;
+	struct nl_msg_buff *msgbuff;
+	struct nl_socket *nlsk;
+	int ret;
+
+	nlctx->cmd = "-L";
+	nlctx->argp = ctx->argp;
+	nlctx->argc = ctx->argc;
+	nlctx->devname = ctx->devname;
+	nlsk = nlctx->ethnl_socket;
+	msgbuff = &nlsk->msgbuff;
+
+	ret = msg_init(nlctx, msgbuff, ETHTOOL_MSG_CHANNELS_SET,
+		       NLM_F_REQUEST | NLM_F_ACK);
+	if (ret < 0)
+		return 2;
+	if (ethnla_fill_header(msgbuff, ETHTOOL_A_CHANNELS_HEADER,
+			       ctx->devname, 0))
+		return -EMSGSIZE;
+
+	ret = nl_parser(nlctx, schannels_params, NULL, PARSER_GROUP_NONE);
+	if (ret < 0)
+		return 1;
+
+	ret = nlsock_sendmsg(nlsk, NULL);
+	if (ret < 0)
+		return 1;
+	ret = nlsock_process_reply(nlsk, nomsg_reply_cb, nlctx);
+	if (ret == 0)
+		return 0;
+	else
+		return nlctx->exit_code ?: 1;
+}
diff --git a/netlink/extapi.h b/netlink/extapi.h
index 8dc18d13a9f9..df6950e8205a 100644
--- a/netlink/extapi.h
+++ b/netlink/extapi.h
@@ -25,6 +25,7 @@ int nl_sprivflags(struct cmd_context *ctx);
 int nl_gring(struct cmd_context *ctx);
 int nl_sring(struct cmd_context *ctx);
 int nl_gchannels(struct cmd_context *ctx);
+int nl_schannels(struct cmd_context *ctx);
 int nl_monitor(struct cmd_context *ctx);
 
 void nl_monitor_usage(void);
@@ -54,6 +55,7 @@ static inline void nl_monitor_usage(void)
 #define nl_gring		NULL
 #define nl_sring		NULL
 #define nl_gchannels		NULL
+#define nl_schannels		NULL
 
 #endif /* ETHTOOL_ENABLE_NETLINK */
 
-- 
2.25.1

